backend/src/routes/getOtp.ts

```diff
@@ -0,0 +1,97 @@
+import { otp, registration } from '@/db/schema';
+import { sendMessage } from '@/utils/sms';
+import { desc, eq } from 'drizzle-orm';
+import type { FastifyInstance } from 'fastify';
+import otpGenerator from 'otp-generator';
+import * as schema from '../db/schema';
+
+interface GetOtpRequestBody {
+  phone_number: string;
+}
+
+// OTP validity duration : 15 mins
+const OTP_VALIDITY_TIME = 900;
+
+// as + symbol is not taken into query params
+// - + will be added after the number is processed
+export function getOtp(fastify: FastifyInstance) {
+  fastify.post<{ Body: GetOtpRequestBody }>(
+    '/get_otp',
+    {
+      schema: {
+        body: {
+          type: 'object',
+          required: ['phone_number'],
+          properties: {
+            phone_number: { type: 'string', pattern: '^\\+[1-9]\\d{1,14}$' },
+          },
+        },
+      },
+    },
+    async (request, reply) => {
+      try {
+        const { phone_number } = request.body as { phone_number: string };
+
+        // validating if phone number exists in db
+        const record_phone_number = await fastify.db.query.registration
+          .findFirst({
+            where: eq(registration.phone_number, phone_number),
+          })
+          .execute();
+
+        if (!record_phone_number) {
+          return reply.code(500).send({ message: 'No record exists with current phone number' });
+        }
+
+        const record = await fastify.db.query.otp
+          .findFirst({
+            where: eq(otp.phone_number, phone_number),
+            orderBy: [desc(otp.created_at)],
+          })
+          .execute();
+
+        if (record) {
+          const time_added = Date.parse(record.created_at) / 1000;
+          const current_time = Math.floor(Date.now() / 1000);
+
+          // Checking if otp already requested
+          // - if time_diff <= 15 mins or otp is used: deny new otp
+          // - else send new otp
+          if (current_time - time_added <= OTP_VALIDITY_TIME) {
+            return reply.code(500).send({ message: 'You have already requested the OTP' });
+          }
+        }
+
+        const otp_gen = generateOtp();
+
+        const send_msg_res = await sendMessage(otp_gen, phone_number);
+        if (!send_msg_res) {
+          fastify.log.error(`Error sending message to phone number : ${phone_number}`);
```

Comment:
You should not log PII.

Reply:
I can see it being useful to know the country code and the type of phone number, so I'd log the first 4 or 6 characters only.

Reply:
done